ledger/participant-integration-api/src/main/scala/platform/apiserver/execution/LedgerTimeAwareCommandExecutor.scala

```diff
@@ -66,39 +78,42 @@ private[apiserver] final class LedgerTimeAwareCommandExecutor(
           else
             contractStore
               .lookupMaximumLedgerTime(usedContractIds)
-              .flatMap { maxUsedTime =>
-                if (maxUsedTime.forall(_ <= commands.commands.ledgerEffectiveTime)) {
-                  Future.successful(Right(cer))
-                } else if (!cer.dependsOnLedgerTime) {
-                  logger.debug(
-                    s"Advancing ledger effective time for the output from ${commands.commands.ledgerEffectiveTime} to $maxUsedTime"
-                  )
-                  Future.successful(Right(advanceOutputTime(cer, maxUsedTime)))
-                } else if (retriesLeft > 0) {
-                  metrics.daml.execution.retry.mark()
-                  logger.debug(
-                    s"Restarting the computation with new ledger effective time $maxUsedTime"
+              .flatMap {
+                case Left(missingContracts) =>
+                  logger.info(
+                    s"Did not find the ledger time for some contracts. This can happen if there is contention on contracts used by the transaction. Missing contracts: ${missingContracts
+                      .mkString(", ")}."
                   )
-                  val advancedCommands = advanceInputTime(commands, maxUsedTime)
-                  loop(advancedCommands, submissionSeed, ledgerConfiguration, retriesLeft - 1)
-                } else {
-                  Future.successful(Left(ErrorCause.LedgerTime(maxRetries)))
-                }
+                  retryOrError(commands)
```

Comment:
Why are we retrying here? Previously these contracts were available, since interpretation worked. If they some are missing now, how can this not be permanent? 🤔

Reply:
they could come from a contract key assignment which has changed since. on reinterpretation we get the new assignment and things might succeed now

Reply:
I see, thanks

Reply:
Ah, I see that we are retrying on the command interpretation as well. Missed that